pkg/streamingpromql/engine_test.go

```diff
@@ -3266,100 +3274,677 @@ func TestQueryStats(t *testing.T) {
+			expr:                        `dense_series{} + end_series{}`,
+			expectedTotalSamples:        11 + 6,
+			expectedTotalSamplesPerStep: []int64{1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2},
 		},
 		"instant vector selector with NaNs": {
-			expr:                 `nan_series{}`,
-			expectedTotalSamples: 11,
+			expr:                        `nan_series{}`,
+			expectedTotalSamples:        11,
+			expectedTotalSamplesPerStep: []int64{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
 		},
 		"range vector selector with NaNs": {
-			expr:                 `sum_over_time(nan_series{}[1m])`,
-			expectedTotalSamples: 11,
+			expr:                        `sum_over_time(nan_series{}[1m])`,
+			expectedTotalSamples:        11,
+			expectedTotalSamplesPerStep: []int64{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
 		},
 		"instant vector selector with native histograms": {
-			expr:                 `native_histogram_series{}`,
-			expectedTotalSamples: 78,
+			expr:                        `native_histogram_series{}`,
+			expectedTotalSamples:        78,
+			expectedTotalSamplesPerStep: []int64{13, 13, 13, 13, 13, 13, 0, 0, 0, 0, 0},
 		},
 		"range vector selector with native histograms": {
-			expr:                 `sum_over_time(native_histogram_series{}[1m])`,
-			expectedTotalSamples: 26,
+			expr:                        `sum_over_time(native_histogram_series{}[1m])`,
+			expectedTotalSamples:        26,
+			expectedTotalSamplesPerStep: []int64{13, 13, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		},
+		"range vector selector with @ modifier": {
+			expr:                        `sum_over_time(dense_series{}[2m] @ 300)`,
+			expectedTotalSamples:        22, // each step selects 2 points at T=300 over query range
+			expectedTotalSamplesPerStep: []int64{2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2},
+		},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			prometheusSamplesStats := runQueryAndGetSamplesStats(t, prometheusEngine, testCase.expr, testCase.isInstantQuery)
+			require.Equal(t, testCase.expectedTotalSamples, prometheusSamplesStats.TotalSamples, "invalid test case: expected total samples does not match value from Prometheus' engine")
+			require.Equal(t, testCase.expectedTotalSamplesPerStep, prometheusSamplesStats.TotalSamplesPerStep, "invalid test case: expected per stepsamples does not match value from Prometheus' engine")
+
+			mimirSamplesStats := runQueryAndGetSamplesStats(t, mimirEngine, testCase.expr, testCase.isInstantQuery)
+			require.Equal(t, testCase.expectedTotalSamples, mimirSamplesStats.TotalSamples)
+			require.Equal(t, testCase.expectedTotalSamplesPerStep, mimirSamplesStats.TotalSamplesPerStep)
+
+			mimirSamplesStatsWithPlanning := runQueryAndGetSamplesStats(t, mimirEngineWithPlanning, testCase.expr, testCase.isInstantQuery)
+			require.Equal(t, testCase.expectedTotalSamples, mimirSamplesStatsWithPlanning.TotalSamples)
+			require.Equal(t, testCase.expectedTotalSamplesPerStep, mimirSamplesStatsWithPlanning.TotalSamplesPerStep)
```

Comment:
[nit] Might be good to add some messages to these assertions to clarify which are with planning enabled and which are without.